### PR TITLE
Code Refactor in Cache Manager

### PR DESCRIPTION
**Pull request by linhong-intel**
## What changes were proposed in this pull request?

1. Remove use of ChunkedByteBuffer to decoupling from Spark BlockMananger
2. Use Oap MemoryManager to allocate off-heap memory
3. Fix some bugs

## How was this patch tested?
Unit test passed.

### CODE REVIEW

**Comment by linhong-intel** [comment]
Each commit is a sub feature. I can divide this PR into small ones if needed.

**Comment by linhong-intel** [comment]
Data type transition:
read data from file: `Array[Byte]`
Allocate off-heap memory for data: `DirectByteBuffer(address: Long)`
Put data into Spark Cache: `ChunkedByteBuffer(chunks: Array[ByteBuffer])`
Get data from Spark Cache: `FiberData(memory: MemoryBlock(obj: null, address: Long))`
Data can't be cached: `FiberData(memory: MemoryBlock(obj: Array[Byte], offset: Long))`

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala [hunk 1]
```diff
@@ -64,8 +104,12 @@ object FiberCacheManager extends Logging {
 
   private val dataFileIdMap = new TimeStampedHashMap[String, DataFile](updateTimeStampOnGet = true)
 
-  private def toByteBuffer(buf: Array[Byte]): ChunkedByteBuffer = {
-    new ChunkedByteBuffer(ByteBuffer.wrap(buf))
+  private def toChunkedByteBuffer(buf: Array[Byte]): ChunkedByteBuffer = {
+    val allocator = MemoryManager.allocateDirectBuffer _
+    val byteBuffer = allocator(buf.length)
+    byteBuffer.put(buf)
+    byteBuffer.flip()
```
Comment (adrian-wang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala]: do we need flip here?

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/MemoryManager.scala [hunk 2]
```diff
@@ -17,38 +17,14 @@
...
-import java.util.concurrent.atomic.AtomicLong
+import java.nio.ByteBuffer
 
 import org.apache.spark.internal.Logging
-import org.apache.spark.unsafe.memory.{MemoryAllocator, MemoryBlock}
-
-
-// TODO: make it an alias of MemoryBlock
-trait FiberCache {
-  def fiberData: MemoryBlock
```
Comment (chenghao-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/MemoryManager.scala]:
```scala
trait FiberCache {
  def fiberData: MemoryBlock
  def getInt(offset: Int): Int = {
    assert(offset + IntegerType.defaultSize < fiberData.length) 
    Platform.getInt(fiberData.obj, fiberData.offset + offset)
  }
  def getDouble(offset: Int): Double = {
    assert(offset + DoubleType.defaultSize < fiberData.length) 
    Platform.getDouble(fiberData.obj, fiberData.offset + offset)
  }
...
}

```

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitMapScanner.scala [hunk 3]
```diff
@@ -101,11 +97,8 @@ private[oap] case class BitMapScanner(idxMeta: IndexMeta) extends IndexScanner(i
     bmFooterBuffer
   }
 
-  private def readBmFooterFromCache(cr: CacheResult): Unit = {
-    val (baseObj, baseOffset): (AnyRef, Long) = cr.buffer.chunks.head match {
-      case db: DirectBuffer => (null, db.address())
-      case _ => (cr.buffer.toArray, Platform.BYTE_ARRAY_OFFSET)
-    }
+  private def readBmFooterFromCache(data: FiberData): Unit = {
```
Comment (chenghao-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitMapScanner.scala]: Should be `data: FiberCache`

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitMapScanner.scala [hunk 4]
```diff
@@ -101,11 +97,8 @@ private[oap] case class BitMapScanner(idxMeta: IndexMeta) extends IndexScanner(i
...
+    val (baseObj, baseOffset): (AnyRef, Long) = (data.baseObj, data.baseOffset)
     bmUniqueKeyListTotalSize = Platform.getInt(baseObj, baseOffset)
```
Comment (chenghao-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitMapScanner.scala]: Call the `data.getInt(xx)` instead.

**Comment by linhong-intel** [comment]
After offline sync, I will upload another implementation. related reviews will be address in another PR.
So close this one.

**Pull request closed without merging**